### ISSUE TITLE
transducer_loss is negative

### ISSUE THREAD

**Issue by Honghe** [TensorSpeech/TensorFlowASR#32]
Environment:
- Ubuntu 20.04
- TensorFlow 2.3
- TensorFlowASR main

Command:
```
python examples/conformer/train_subword_conformer.py --tbs 8 --ebs 8 --mxp --devices 0 --cache --subwords ./output/librispeed \
--subwords_corpus \
/home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/train-clean-100/transcripts.tsv \
/home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/test-clean/transcripts.tsv \
/home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/dev-clean/transcripts.tsv \
/home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/dev-other/transcripts.tsv 
```

Log:
```
Model: "conformer"
________________________________________________________________________________________________________________________
Layer (type)                                          Output Shape                                    Param #           
========================================================================================================================
conformer_encoder (ConformerEncoder)                  (None, None, 144)                               8710848           
________________________________________________________________________________________________________________________
conformer_prediction (TransducerPrediction)           (None, None, 320)                               1151040           
________________________________________________________________________________________________________________________
conformer_joint (TransducerJoint)                     (None, None, None, 1031)                        479751            
========================================================================================================================
Total params: 10,341,639
Trainable params: 10,337,031
Non-trainable params: 4,608
________________________________________________________________________________________________________________________
Reading /home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/train-clean-100/transcripts.tsv ...
Reading /home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/dev-clean/transcripts.tsv ...
Reading /home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/dev-other/transcripts.tsv ...
[Train] |                    | 0/71340 [00:00<?, ?batch/s]2020-10-21 10:26:47.344713: W tensorflow/stream_executor/gpu/asm_compiler.cc:81] Running ptxas --version returned 256
2020-10-21 10:26:47.417289: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] Internal: ptxas exited with non-zero error code 256, output: 
Relying on driver to perform ptx compilation. 
Modify $PATH to customize ptxas location.
This message will be only logged once.
[Train] [Epoch 1/20] |                    | 45/71340 [02:59<71:52:55,  3.63s/batch, transducer_loss=-298.91907] 
```

**Comment by usimarit** [comment]
Hi @Honghe, the cause is wrong installation of rnnt loss, remember to export CUDA_HOME before running `scripts/install_rnnt_loss.sh`

**Comment by Honghe** [comment]
After rebuilt warprnnt_tensorflow with CUDA, it seems to be running correctly.
```
[Train] [Epoch 1/20] |                    | 56/142680 [00:42<6:16:25,  6.31batch/s, transducer_loss=2464.3044]  
```
But it will consume 6 hours to train one epoch, is this normal?
Environment: i7 cpu, one 2080Ti GPU.

**Comment by usimarit** [comment]
@Honghe in your case, it's the total of 6 hours for 20 epochs

**Comment by Honghe** [comment]
@usimarit ` --tbs 8 --ebs 8` makes TensorFlow CUDA OOM after a few steps, is there memory leaking somewhere?
Batch size 8 is a little small, how to effectively increase batch size?
```
python examples/conformer/train_subword_conformer.py --tbs 8 --ebs 8 --mxp --devices 0 --cache --subwords ./output/librispeed \
--subwords_corpus \
/home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/train-clean-100/transcripts.tsv \
/home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/test-clean/transcripts.tsv \
/home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/dev-clean/transcripts.tsv \
/home/ubuntu/Data/LibriSpeechConformer/LibriSpeech/dev-other/transcripts.tsv 
```

**Comment by usimarit** [comment]
@Honghe I'm working on gradient accumulation to increase batch size. Technically, when training ASR, batch size small is normal, there's no memory leaks because this model requires lots of memory (due to multi-head self attention), regardless its small number of parameters.

**Comment by OmarHory** [comment]
Just to reassure, in his case above 6 hours for 20 epochs?  I haven't seen such thing like this before, usually the time is per epoch.

**Comment by namdn** [comment]
I already install rnnt loss by `export CUDA_HOME=/usr/local/cuda` and `scripts/install_rnnt_loss.sh` but the `transducer_loss` is still negative. Any one can help me. Thanks 
![tensorboard](https://user-images.githubusercontent.com/5409672/112276448-56e7d280-8cb3-11eb-9455-2155dd72a7d5.png)

**Comment by usimarit** [comment]
@namdn Please use rnnt_loss in TF, the warp-transducer is deprecated

**Comment by namdn** [comment]
@usimarit how can I run without `warp-transducer`.

**Comment by namdn** [comment]
I mean i do the same as you say in Readme. `For training Transducer Models with RNNT Loss in TF, make sure that warp-transducer is not installed (by simply run pip3 uninstall warprnnt-tensorflow)`. But the `transducer_loss` is still negative.

**Comment by usimarit** [comment]
@namdn then please check your dataset and vocabulary file, make sure the vocabulary file covers all of characters or subwords in your dataset.

**Comment by usimarit** [comment]
@namdn if you are using subwords and tfrecords, make sure you pass the same subword file that used to generate tfrecords.

**Comment by namdn** [comment]
Now i am using `CharFeaturizer` as `text_featureizers` and using `script train_conformer.py` for trainning. In `config.yml` the value for `decoder_config.vocabulary` is `language.characters` file? I configurate it to my language characters file and the value is still negative. Any other suggestion?

**Comment by usimarit** [comment]
@namdn can you try script `train_keras_conformer.py`?

**Comment by namdn** [comment]
thank you very much. It worked for me.
